src/v/datalake/protobuf_to_arrow_converter.cc

```diff
@@ -0,0 +1,134 @@
+      serialized_message);
+    if (message == nullptr) {
+        return arrow_converter_status::parse_error;
+    }
+
+    if (!_struct_converter->add_top_level_message(message.get()).ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+arrow_converter_status proto_to_arrow_converter::finish_batch() {
+    if (!_struct_converter->finish_batch().ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+std::shared_ptr<arrow::Table> proto_to_arrow_converter::build_table() {
+    auto table_result = arrow::Table::FromChunkedStructArray(
+      _struct_converter->finish());
+    if (table_result.ok()) {
+        return table_result.ValueUnsafe();
+    } else {
+        return nullptr;
+    }
+}
+std::vector<std::shared_ptr<arrow::Field>>
+proto_to_arrow_converter::build_field_vec() {
+    return _struct_converter->get_field_vector();
+}
+std::shared_ptr<arrow::Schema> proto_to_arrow_converter::build_schema() {
+    return arrow::schema(_struct_converter->get_field_vector());
+}
+
+void proto_to_arrow_converter::initialize_protobuf_schema(
+  const ss::sstring& schema) {
+    google::protobuf::io::ArrayInputStream proto_input_stream(
+      schema.c_str(), static_cast<int>(schema.size()));
+    google::protobuf::io::Tokenizer tokenizer(
+      &proto_input_stream, &error_collector);
+
+    google::protobuf::compiler::Parser parser;
+    parser.RecordErrorsTo(&error_collector);
+    if (!parser.Parse(&tokenizer, &_file_descriptor_proto)) {
+        throw initialization_error("Could not parse protobuf schema");
+    }
+
+    if (!_file_descriptor_proto.has_name()) {
+        _file_descriptor_proto.set_name("default_message_name");
+    }
+
+    _file_desc = _protobuf_descriptor_pool.BuildFile(_file_descriptor_proto);
+    if (_file_desc == nullptr) {
+        throw initialization_error("Could not build descriptor pool");
+    }
+}
+bool proto_to_arrow_converter::initialize_struct_converter() {
+    using namespace detail;
+    namespace pb = google::protobuf;
+
+    const pb::Descriptor* message_desc = message_descriptor();
+    if (message_desc == nullptr) {
+        return false;
+    }
+
+    _struct_converter = std::make_unique<detail::proto_to_arrow_struct>(
+      message_desc);
+
+    return true;
+}
+std::unique_ptr<google::protobuf::Message>
+proto_to_arrow_converter::parse_message(const ss::sstring& message) {
+    // Get the message descriptor
+    const google::protobuf::Descriptor* message_desc = message_descriptor();
+    if (message_desc == nullptr) {
+        return nullptr;
+    }
+
+    const google::protobuf::Message* prototype_msg = _factory.GetPrototype(
+      message_desc);
+    if (prototype_msg == nullptr) {
+        return nullptr;
+    }
+
+    std::unique_ptr<google::protobuf::Message> mutable_msg
+      = std::unique_ptr<google::protobuf::Message>(prototype_msg->New());
+    if (mutable_msg == nullptr) {
```

Comment:
what are the failure modes here? should this throw bad_alloc? is there some error messages coming out of protobuf library that we can log? similar to above where the prototype wasn't found.